### PR TITLE
ENT-3570: Adjust config for event pipeline

### PR DESCRIPTION
**Pull request by dmckoske**
ENT-3570: Adjust config for event pipeline.  

This change delays the latency window by an hour, with the goal of making sure all metering tasks are finished before the tally starts.

### CODE REVIEW

File: src/main/java/org/candlepin/subscriptions/metering/service/prometheus/MetricProperties.java [hunk 1]
```diff
@@ -70,68 +80,9 @@
...
-        return backOffMultiplier;
-    }
-
-    public void setBackOffMultiplier(double backOffMultiplier) {
-        this.backOffMultiplier = backOffMultiplier;
-    }
+    /**
+     * Batch size to use while persisting events.
+     */
+    private int eventBatchSize = 1000;
```
Comment (dmckoske) [inline src/main/java/org/candlepin/subscriptions/metering/service/prometheus/MetricProperties.java:87]: I'm confused by a recent merge conflict.  Is eventBatchSize supposed to exist here, or did I add it back by mistake?

Comment (kahowell) [reply src/main/java/org/candlepin/subscriptions/metering/service/prometheus/MetricProperties.java:87]: Looks like accidental reintroduction.

File: src/main/java/org/candlepin/subscriptions/metering/service/prometheus/MetricProperties.java [hunk 2]
```diff
@@ -44,6 +49,11 @@
      */
     private int rangeInMinutes = 60;
 
+    /**
+     * Latency offset: how many minutes in the past you would like the metering job latency (window) to occur.
+     */
+    private int offsetInMinutes = 60;
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/metering/service/prometheus/MetricProperties.java:55]: I'd like to make sure this is consistent with the existing `prometheusLatencyDuration` property.

- [ ] Instead of defining this here, can you please move to ApplicationProperties for consistency?
- [ ] Can you please define as a `Duration`?
- [ ] Can you please rename to reflect that this offset applies to the range of time for the hourly tally (e.g. `hourlyTallyOffset`?)
- [ ] Can you please leave a comment about why this offset is needed?
- [ ] Can you please define an environment variable for this in `src/main/resources/application.yaml` ?

File: src/main/java/org/candlepin/subscriptions/metering/job/OpenShiftMeteringJob.java [hunk 3]
```diff
@@ -59,13 +59,17 @@ public OpenShiftMeteringJob(PrometheusMetricsTaskManager tasks, ApplicationClock
     public void run() {
         Duration latency = appProps.getPrometheusLatencyDuration();
         int range = metricProperties.getOpenshift().getRangeInMinutes();
-        OffsetDateTime startDate = clock.startOfHour(clock.now().minus(latency).minusMinutes(range));
+        int offset = metricProperties.getOpenshift().getOffsetInMinutes();
+
+        OffsetDateTime startDate = clock.startOfHour(clock.now().minus(latency)
+            .minusMinutes(range).minusMinutes(offset));
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/metering/job/OpenShiftMeteringJob.java:65]: This is precisely the kind of change we need, but we need to apply it instead in `org.candlepin.subscriptions.tally.job.CaptureSnapshotsTaskManager#updateHourlySnapshotsForAllAccounts`

Rationale follows;

Given:
- current `develop` version of the code
- a running instance of `OpenShiftMeteringJob` at ~1am UTC.
- a running instance of `CaptureMetricsSnapshotTask` (scheduled by `CaptureSnapshotsTaskManager` at ~1am UTC)
- default configurations

then:
- the range for the OpenShiftMeteringJob would be (`1am - 4 hours - 1 hour` through `1am - 4 hours`) -> (8pm to 9pm)
- the range for the CaptureMetricsSnapshotTask would be (`1 am - 4 hours - 24 hours` through `1am - 4 hours`) -> (9pm to 9pm)

In other words, CaptureMetricsSnapshotTask will currently operate on an overlapping hour with the metering job.

Please let me know if the above makes sense and is accurate (some smoketesting might be helpful).

**Review by kahowell — CHANGES_REQUESTED** [review]

File: src/main/java/org/candlepin/subscriptions/metering/job/OpenShiftMeteringJob.java (hunk 3 above)
Comment (dmckoske) [reply src/main/java/org/candlepin/subscriptions/metering/job/OpenShiftMeteringJob.java:65]: That makes sense.  I will add this to the card.  PR forthcoming.

File: src/test/java/org/candlepin/subscriptions/tally/job/CaptureSnapshotsTaskManagerTest.java [hunk 4]
```diff
@@ -135,6 +140,23 @@ void ensureNoUpdatesWhenAccountListCanNotBeRetreived() throws Exception {
         verify(queue, never()).enqueue(any());
     }
 
+    /**
+     * .
+     *
```
Comment (dmckoske) [inline src/test/java/org/candlepin/subscriptions/tally/job/CaptureSnapshotsTaskManagerTest.java:145]: Oversight.  Will correct.

File: src/main/java/org/candlepin/subscriptions/ApplicationProperties.java [hunk 5]
```diff
@@ -165,4 +165,12 @@
      */
     private Duration metricLookupRangeDuration = Duration.ofHours(24L);
 
+    /**
+     * Latency offset: how many minutes in the past you would like the metering job latency (window) to occur.
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/ApplicationProperties.java:169]:
```suggestion
     * Latency offset: how far back to offset the hourly tally window.
```
?

File: src/test/java/org/candlepin/subscriptions/tally/job/CaptureSnapshotsTaskManagerTest.java [hunk 6]
```diff
@@ -145,6 +168,26 @@ private TaskDescriptor createDescriptor(List<String> accounts) {
...
+            .truncatedTo(ChronoUnit.HOURS), prometheusLatencyDuration);
+        OffsetDateTime startDateTime = endDateTime.minus(metricRange).minus(hourlyTallyOffsetMinutes);
+
+        return TaskDescriptor.builder(TaskType.UPDATE_HOURLY_SNAPSHOTS, taskQueueProperties.getTopic())
+            .setSingleValuedArg("accountNumber", accountNumber)
+            .setSingleValuedArg("startDateTime", startDateTime.toString())
+            .setSingleValuedArg("endDateTime", endDateTime.toString())
+            .build();
+    }
+
```
Comment (kahowell) [inline src/test/java/org/candlepin/subscriptions/tally/job/CaptureSnapshotsTaskManagerTest.java:190]: Rather than repeat the logic from the `CaptureSnapshotsTaskManager`, I'd rather see real numbers in here. Specifically, I'd rather see this method take `startDateTime` and `endDateTime` as parameters. Then the test will ensure that we get the expected values for `startDateTime` and `endDateTime`.

It would help to use `@ContextConfiguration(classes = FixedClockConfiguration.class)`

Note this will actually require a tiny tweak to `CaptureSnapshotsTaskManager`: you'll need to inject the `ApplicationClock` and change `OffsetDateTime.now()` to `OffsetDateTime.now(applicationClock.getClock())` (as seen in `src/main/java/org/candlepin/subscriptions/retention/TallyRetentionPolicy.java`.

**Review by kahowell — CHANGES_REQUESTED** [review]
The logic change is now spot on, and the comments help a lot for understanding! I left a couple of requests inline. FYI, I haven't tested this locally yet, I may have some more feedback after that.

File: src/main/java/org/candlepin/subscriptions/ApplicationProperties.java [hunk 7]
```diff
@@ -165,4 +165,12 @@
...
+     *
+     * The offset is subtracted from the beginning and ending times of the latency window, to delay the entire
+     * processing window.  This ensures more metering tasks finish and report their totals before tallying
+     * begins.
+     */
+    private Duration hourlyTallyOffsetMinutes = Duration.ofMinutes(60L);
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/ApplicationProperties.java:175]: Please refactor the name:

s/hourlyTallyOffsetMinutes/hourlyTallyOffset/

Until I did this locally, the env var did not work as expected.

File: src/main/java/org/candlepin/subscriptions/ApplicationProperties.java (hunk 5 above)
Comment (dmckoske) [reply src/main/java/org/candlepin/subscriptions/ApplicationProperties.java:169]: I sill like mine better, but okay.

File: src/main/java/org/candlepin/subscriptions/ApplicationProperties.java (hunk 7 above)
Comment (dmckoske) [reply src/main/java/org/candlepin/subscriptions/ApplicationProperties.java:175]: Ah, crud.  My fault.  I will change this.

File: src/main/java/org/candlepin/subscriptions/ApplicationProperties.java (hunk 5 above)
Comment (kahowell) [reply src/main/java/org/candlepin/subscriptions/ApplicationProperties.java:169]: Open to other opinions too, @mstead @lindseyburnett @awood preferences? or suggestions?

File: src/main/java/org/candlepin/subscriptions/tally/job/CaptureSnapshotsTaskManager.java [hunk 8]
```diff
@@ -137,13 +137,15 @@ public void updateHourlySnapshotsForAllAccounts() {
...
             Duration prometheusLatencyDuration = appProperties.getPrometheusLatencyDuration();
+            Duration hourlyTallyOffsetMinutes = appProperties.getHourlyTallyOffsetMinutes();
 
-            OffsetDateTime endTime = adjustTimeForLatency(OffsetDateTime.now().truncatedTo(ChronoUnit.HOURS),
-                prometheusLatencyDuration);
-            OffsetDateTime startTime = endTime.minus(metricRange);
+            OffsetDateTime endDateTime = adjustTimeForLatency(OffsetDateTime.now()
+                .minus(hourlyTallyOffsetMinutes).truncatedTo(ChronoUnit.HOURS), prometheusLatencyDuration);
+
+            OffsetDateTime startDateTime = endDateTime.minus(metricRange).minus(hourlyTallyOffsetMinutes);
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/tally/job/CaptureSnapshotsTaskManager.java:145]: @dmckoske found one more tiny thing - since the end date time already has the offset applied, don't need to apply again to the `startTime`.

```suggestion
            OffsetDateTime startDateTime = endDateTime.minus(metricRange);
```

FWIW, I've been looking at the output of `HOURLY_TALLY_OFFSET=3h SPRING_PROFILES_ACTIVE=capture-snapshots ./gradlew bootRun` to verify.

File: src/test/java/org/candlepin/subscriptions/tally/job/CaptureSnapshotsTaskManagerTest.java (hunk 6 above)
Comment (dmckoske) [reply src/test/java/org/candlepin/subscriptions/tally/job/CaptureSnapshotsTaskManagerTest.java:190]: One concern I have about this is this change puts this test out of sync with the way other tests are done, and with the way the CaptureSnapshotsTaskmanager.update... method is implemented.  I don't like copy / pasting code either, sure.  How literally does the test need to mimic / emulate the way the tested method works?

Comment (dmckoske) [reply src/test/java/org/candlepin/subscriptions/tally/job/CaptureSnapshotsTaskManagerTest.java:190]: I thought the test already checked the values.  The error I saw while sorting out the test suggests this.

File: src/main/java/org/candlepin/subscriptions/tally/job/CaptureSnapshotsTaskManager.java (hunk 8 above)
Comment (dmckoske) [reply src/main/java/org/candlepin/subscriptions/tally/job/CaptureSnapshotsTaskManager.java:145]: Yeah, sorry about that.  I see the calculation now.

File: src/main/java/org/candlepin/subscriptions/ApplicationProperties.java (hunk 5 above)
Comment (dmckoske) [reply src/main/java/org/candlepin/subscriptions/ApplicationProperties.java:169]: Bueller?

File: src/test/java/org/candlepin/subscriptions/tally/job/CaptureSnapshotsTaskManagerTest.java [hunk 9]
```diff
@@ -135,6 +148,33 @@ void ensureNoUpdatesWhenAccountListCanNotBeRetreived() throws Exception {
...
+        OffsetDateTime endDateTime = adjustTimeForLatency(OffsetDateTime.now(applicationClock.getClock())
+            .minus(hourlyTallyOffsetMinutes).truncatedTo(ChronoUnit.HOURS), prometheusLatencyDuration);
+        OffsetDateTime startDateTime = endDateTime.minus(metricRange);
+
+        manager.updateHourlySnapshotsForAllAccounts();
+
+        expectedAccounts.forEach(accountNumber -> {
+            verify(queue, times(1)).enqueue(createHourlyDescriptor(accountNumber,
+                startDateTime.toString(), endDateTime.toString()));
+        });
```
Comment (kahowell) [inline src/test/java/org/candlepin/subscriptions/tally/job/CaptureSnapshotsTaskManagerTest.java:175]: This is an example of how I'd prefer this test be implemented (there are other ways to do this I'd be fine with too, e.g. org.mockito.ArgumentCaptor). I realize it is different than other tests, but this intentional. I'm concerned with making sure we test the math is correct rather than repeat the math in the test. Otherwise a flaw in the code that does all the arithmetic will occur in both the test and in the code being tested :-)

```suggestion
        expectedAccounts.forEach(accountNumber -> {
            verify(queue, times(1)).enqueue(
                TaskDescriptor.builder(TaskType.UPDATE_HOURLY_SNAPSHOTS, taskQueueProperties.getTopic())
                .setSingleValuedArg("accountNumber", accountNumber)
                // 2019-05-24T12:35Z truncated to top of the hour - 4 hours prometheus latency - 1 hour
                // tally latency - 24 hour metric range
                .setSingleValuedArg("startDateTime", "2019-05-23T07:00Z")
                .setSingleValuedArg("endDateTime", "2019-05-24T07:00Z").build());
        });
```

**Review by kahowell — CHANGES_REQUESTED** [review]

File: src/main/java/org/candlepin/subscriptions/ApplicationProperties.java [hunk 10]
```diff
@@ -165,4 +165,12 @@
      */
     private Duration metricLookupRangeDuration = Duration.ofHours(24L);
 
+    /**
+     * Latency offset: how back to set the hourly tally window.
```
Comment (mstead) [inline src/main/java/org/candlepin/subscriptions/ApplicationProperties.java:169]: 'how _far_ back'

File: src/main/java/org/candlepin/subscriptions/tally/job/CaptureSnapshotsTaskManager.java [hunk 11]
```diff
@@ -137,13 +142,14 @@ public void updateHourlySnapshotsForAllAccounts() {
 
             Duration metricRange = appProperties.getMetricLookupRangeDuration();
             Duration prometheusLatencyDuration = appProperties.getPrometheusLatencyDuration();
+            Duration hourlyTallyOffsetMinutes = appProperties.getHourlyTallyOffset();
 
-            OffsetDateTime endTime = adjustTimeForLatency(OffsetDateTime.now().truncatedTo(ChronoUnit.HOURS),
-                prometheusLatencyDuration);
-            OffsetDateTime startTime = endTime.minus(metricRange);
+            OffsetDateTime endDateTime = adjustTimeForLatency(OffsetDateTime.now(applicationClock.getClock())
+                .minus(hourlyTallyOffsetMinutes).truncatedTo(ChronoUnit.HOURS), prometheusLatencyDuration);
```
Comment (mstead) [inline src/main/java/org/candlepin/subscriptions/tally/job/CaptureSnapshotsTaskManager.java:148]: Change:
`
OffsetDateTime.now(applicationClock.getClock())
`
To:
`applicationClock.now()`

Could even take that as far as the following, but I'll leave that up to you.
```java
applicationClock.startOfHour(applicationClock.now().minus(hourlyTallyOffsetMinutes))
```

**Review by mstead — CHANGES_REQUESTED** [review]
Couple very minor things. Nothing more to add than what Kevin has already mentioned.

File: src/main/java/org/candlepin/subscriptions/ApplicationProperties.java (hunk 10 above)
Comment (dmckoske) [reply src/main/java/org/candlepin/subscriptions/ApplicationProperties.java:169]: Good catch.  I'm having trouble English.

File: src/main/java/org/candlepin/subscriptions/tally/job/CaptureSnapshotsTaskManager.java (hunk 11 above)
Comment (dmckoske) [reply src/main/java/org/candlepin/subscriptions/tally/job/CaptureSnapshotsTaskManager.java:148]: Done and updated

Comment (kahowell) [reply src/main/java/org/candlepin/subscriptions/tally/job/CaptureSnapshotsTaskManager.java:148]: @dmckoske did you maybe forget to commit this file? Not seeing the change...

Comment (dmckoske) [reply src/main/java/org/candlepin/subscriptions/tally/job/CaptureSnapshotsTaskManager.java:148]: That's odd.  I see the change above.

Comment (dmckoske) [reply src/main/java/org/candlepin/subscriptions/tally/job/CaptureSnapshotsTaskManager.java:148]: Have a look in the files tab.  The change is there.  The quote above must not get updated with subsequent checkins or something.

Comment (kahowell) [reply src/main/java/org/candlepin/subscriptions/tally/job/CaptureSnapshotsTaskManager.java:148]: Weird. Usually it gets marked as outdated... I am good now, if you want to squash.

**Review by kahowell — APPROVED** [review]

**Review by mstead — APPROVED** [review]

**Review by kahowell — APPROVED** [review]
